### PR TITLE
Copy without Lockfiles

### PR DESCRIPTION
**Pull request by tnt-manu**
Initial version, lacks proper tests. Worked on local machine with multiple sample files and multiple source directories on two times "task copy".

### CODE REVIEW

File: chia_tea/copy_plots/Disk.py [hunk 1]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
     return all_filepaths
 
 
-def update_lockfile_count(target_dirs: List[str]) -> Dict[str, int]:
-    """Get the lockfile count for the specified directories
+def is_accessible(fpath):
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:181]:
```suggestion
def is_accessible(fpath: str):
```

File: chia_tea/copy_plots/Disk.py [hunk 2]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
+
+    Returns
+    -------
+    accessible: boolean
+        false if being used by another process
+    """
+    try:
+        with open(fpath, "r+", encoding="utf8"):
+            pass
+    except Exception:
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:196]: Would be nice to limit the exception type. Generally we should note what exceptions can occur (including NAS connection issues)

File: chia_tea/copy_plots/Disk.py [hunk 3]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
-        n_lockfiles = len(lock_files)
-        n_lockfiles_per_disk[target_dir] = n_lockfiles
-    return n_lockfiles_per_disk
+        files_beeing_copied_to_dir = get_files_beingCopied([target_dir], files_copied_completely)
+        number_of_copy_processes = len(files_beeing_copied_to_dir)
+        number_of_copy_processes_per_disk[target_dir] = number_of_copy_processes
+    return number_of_copy_processes_per_disk
+
+
+def get_files_beingCopied(target_dirs: List[str], files_copied_completely: List[str]) -> List[str]:
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:224]:
```suggestion
def get_files_being_copied(target_dirs: List[str], files_copied_completely: List[str]) -> List[str]:
```

File: chia_tea/copy_plots/Disk.py [hunk 4]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
+    """
+    Later
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:226]: someone was tired I guess

File: chia_tea/copy_plots/Disk.py [hunk 5]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
+            if os.path.isfile(folder_path):
+                warn_msg = "Path '{0}' is a file and not a directory."
+            else:
+                warn_msg = "Folder '{0}' does not exist or is not a directory."
+            logger.warning(warn_msg, folder_path)
+            continue
+
+        all_files_to_check = [f for f in os.listdir(folder_path) if isfile(join(folder_path, f))]
+
+        all_files_to_check = [f for f in os.listdir(folder_path) if isfile(join(folder_path, f))]
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:242]:
```suggestion
        all_files_to_check = [f for f in os.listdir(folder_path) if isfile(join(folder_path, f))]
```

File: chia_tea/copy_plots/Disk.py [hunk 6]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
+
+        # remove all files which not have to be cheked
+        for f in files_copied_completely:
+            if f in all_files_to_check:
+                all_files_to_check.remove(f)
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:247]: You are working with Lists here, try using sets. They are made exactly for all that stuff. And use `os.path.samefile` to make sure that relative and absolute filepaths are handled correctly or simply expand all filepaths.

File: chia_tea/copy_plots/Disk.py [hunk 7]
```diff
@@ -141,23 +178,79 @@ def collect_files_from_folders(folder_list: List[str], pattern: str) -> List[str
...
+        number_of_copy_processes = len(files_beeing_copied_to_dir)
+        number_of_copy_processes_per_disk[target_dir] = number_of_copy_processes
+    return number_of_copy_processes_per_disk
+
+
+def get_files_beingCopied(target_dirs: List[str], files_copied_completely: List[str]) -> List[str]:
+    """
+    Later
+    """
+    all_filepaths = []
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:228]: How about renaming this to `filepaths_copied_files`?

File: chia_tea/copy_plots/__main__.py [hunk 8]
```diff
@@ -53,11 +61,7 @@ def main():
...
 
-            # lockfile
-            path_to_lockfile = os.path.join(target_dir, filename.replace(".plot", ".copying"))
-
-            with create_lockfile(path_to_lockfile):
-                successful_copy = copy_file(filepath, target_path)
+            successful_copy = copy_file(filepath, target_path)
 
             duration_secs = time.time() - start
             if successful_copy:
```
Comment (tnt-codie) [inline chia_tea/copy_plots/__main__.py:67]:
```suggestion
            if not successful_copy:
```

**Review by tnt-codie — CHANGES_REQUESTED** [review]
Generally I like the solution a lot. Think this is a decent step forward. My only thought is, the feature is so critical, we should write a few unittests for it first. Since there are a lot of side-effects, we definitely need to patch the os functions writing and reading but that is no problem with `unittest.mock`.

File: chia_tea/copy_plots/Disk.py (hunk 4 above)
Comment (tnt-manu) [reply chia_tea/copy_plots/Disk.py:226]: that was indeed correct

File: chia_tea/copy_plots/Disk.py (hunk 7 above)
Comment (tnt-manu) [reply chia_tea/copy_plots/Disk.py:228]: yes!

File: chia_tea/copy_plots/Disk.py (hunk 2 above)
Comment (tnt-manu) [reply chia_tea/copy_plots/Disk.py:196]: not completely sure about the nas connection error ... implemented OSError and FileNotFoundError :)

File: chia_tea/copy_plots/__main__.py (hunk 8 above)
Comment (tnt-manu) [reply chia_tea/copy_plots/__main__.py:67]: ? sure ?

File: chia_tea/copy_plots/Disk.py (hunk 6 above)
Comment (tnt-manu) [reply chia_tea/copy_plots/Disk.py:247]: oh thats a bigger one .. I think I will do that tomorrow with a fresh head ... so its sets for all file lists ?

File: chia_tea/copy_plots/__main__.py (hunk 8 above)
Comment (tnt-codie) [reply chia_tea/copy_plots/__main__.py:70]: You log failed in case of success?

File: chia_tea/copy_plots/Disk.py [hunk 9]
```diff
@@ -85,79 +125,165 @@ def copy_file(source_path: str, target_path: str) -> bool:
...
+        warn_msg = "File access check: '%s' does not exist."
+        logger.warning(warn_msg, os.path.abspath(fpath))
+        return False
+    except ConnectionResetError:
+        warn_msg = "File access check: Lost connection to network on file: '%s'"
+        logger.warning(warn_msg, os.path.abspath(fpath))
+        return False
+    except OSError:
+        warn_msg = "File access check: Cannot reach file '%s'"
+        logger.warning(warn_msg, os.path.abspath(fpath))
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:217]: No `False` on `OSError`?

File: chia_tea/copy_plots/Disk.py [hunk 10]
```diff
@@ -20,26 +21,65 @@ def filter_least_used_disks(disk_to_lockfile_count: Dict[str, int]) -> List[str]
...
+        Directories in which the file can be copied
+
+    Returns
+    -------
+    files_copied_completely: Set[str]
+        All the files which are not being copied anymore
+    """
+    logger = get_logger(__file__)
+    for folder_path in target_dirs:
+        print("Target:" + folder_path)
```
Comment (tnt-codie) [inline chia_tea/copy_plots/Disk.py:54]: Use the logger we have.

**Comment by codie3611** [comment]
@tnt-manu I'm still overhauling this PR. Wanna use it in the next release. The testing helped with clearing the code but the os side side effects made testing no fun.

**Review by codie3611 — APPROVED** [review]

**Pull request merged**